[Bug]: create collection failed to assign channels

### Is there an existing issue for this?

- [X] I have searched the existing issues

### Environment

```markdown
- Milvus version:
- Deployment mode(standalone or cluster):
- SDK version(e.g. pymilvus v2.0.0rc2):
- OS(Ubuntu or CentOS): 
- CPU/Memory: 
- GPU: 
- Others:
```


### Current Behavior

[2022/09/28 15:48:17.935 +00:00] [INFO] [rootcoord/root_coord.go:688] ["received request to create collection"] [traceID=5ff259332f0adbef] [role=rootcoord] [name=collection_a442vBK8] [msgID=436307238041681940]
panic: runtime error: index out of range [0] with length 0

goroutine 857 [running]:
github.com/milvus-io/milvus/internal/rootcoord.(*createCollectionTask).assignChannels(0xc005e3bb00)
	/go/src/github.com/milvus-io/milvus/internal/rootcoord/create_collection_task.go:136 +0x285
github.com/milvus-io/milvus/internal/rootcoord.(*createCollectionTask).Prepare(0xc005e3bb00, {0xc002ccef10?, 0xc002277f08?})
	/go/src/github.com/milvus-io/milvus/internal/rootcoord/create_collection_task.go:164 +0xcf
github.com/milvus-io/milvus/internal/rootcoord.(*scheduler).execute(0xc002277f90?, {0x3e3de08, 0xc005e3bb00})
	/go/src/github.com/milvus-io/milvus/internal/rootcoord/scheduler.go:55 +0x47
github.com/milvus-io/milvus/internal/rootcoord.(*scheduler).taskLoop(0xc003464900)
	/go/src/github.com/milvus-io/milvus/internal/rootcoord/scheduler.go:70 +0x8a
created by github.com/milvus-io/milvus/internal/rootcoord.(*scheduler).Start
	/go/src/github.com/milvus-io/milvus/internal/rootcoord/scheduler.go:46 +0x6a

### Expected Behavior

_No response_

### Steps To Reproduce

_No response_

### Milvus Log

_No response_

### Anything else?

_No response_

--- a/internal/rootcoord/create_collection_task.go
+++ b/internal/rootcoord/create_collection_task.go
@@ -128,11 +128,15 @@ func (t *createCollectionTask) assignPartitionID() error {
 }
 
 func (t *createCollectionTask) assignChannels() error {
-	vchanNames := make([]string, t.Req.ShardsNum)
+	vchanNames := make([]string, t.Req.GetShardsNum())
 	//physical channel names
-	chanNames := t.core.chanTimeTick.getDmlChannelNames(int(t.Req.ShardsNum))
+	chanNames := t.core.chanTimeTick.getDmlChannelNames(int(t.Req.GetShardsNum()))
 
-	for i := int32(0); i < t.Req.ShardsNum; i++ {
+	if int32(len(chanNames)) < t.Req.GetShardsNum() {
+		return fmt.Errorf("no enough channels, want: %d, got: %d", t.Req.GetShardsNum(), len(chanNames))
+	}
+
+	for i := int32(0); i < t.Req.GetShardsNum(); i++ {
 		vchanNames[i] = fmt.Sprintf("%s_%dv%d", chanNames[i], t.collID, i)
 	}
 	t.channels = collectionChannels{

--- a/internal/rootcoord/create_collection_task_test.go
+++ b/internal/rootcoord/create_collection_task_test.go
@@ -218,10 +218,6 @@ func Test_createCollectionTask_Prepare(t *testing.T) {
 		assert.Error(t, err)
 	})
 
-	t.Run("failed to assign channels", func(t *testing.T) {
-		// TODO: error won't happen here.
-	})
-
 	t.Run("normal case", func(t *testing.T) {
 		defer cleanTestEnv()
 
@@ -251,6 +247,10 @@ func Test_createCollectionTask_Prepare(t *testing.T) {
 				Schema:         marshaledSchema,
 			},
 		}
+		task.Req.ShardsNum = int32(Params.RootCoordCfg.DmlChannelNum + 1) // no enough channels.
+		err = task.Prepare(context.Background())
+		assert.Error(t, err)
+		task.Req.ShardsNum = 1
 		err = task.Prepare(context.Background())
 		assert.NoError(t, err)
 	})

--- a/internal/rootcoord/dml_channels.go
+++ b/internal/rootcoord/dml_channels.go
@@ -180,11 +180,11 @@ func newDmlChannels(ctx context.Context, factory msgstream.Factory, chanNamePref
 }
 
 func (d *dmlChannels) getChannelNames(count int) []string {
+	d.mut.Lock()
+	defer d.mut.Unlock()
 	if count > len(d.channelsHeap) {
 		return nil
 	}
-	d.mut.Lock()
-	defer d.mut.Unlock()
 	// get next count items from heap
 	items := make([]*dmlMsgStream, 0, count)
 	result := make([]string, 0, count)

--- a/internal/rootcoord/dml_channels_test.go
+++ b/internal/rootcoord/dml_channels_test.go
@@ -149,6 +149,9 @@ func TestDmlChannels(t *testing.T) {
 	dml.addChannels(chans1...)
 	assert.Equal(t, 2, dml.getChannelNum())
 
+	chans2 := dml.getChannelNames(totalDmlChannelNum + 1)
+	assert.Nil(t, chans2)
+
 	dml.removeChannels(chans1...)
 	assert.Equal(t, 2, dml.getChannelNum())
 

